firestore/bulkwriter.go

```diff
@@ -0,0 +1,521 @@
+package firestore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	vkit "cloud.google.com/go/firestore/apiv1"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+)
+
+const (
+	// MaxBatchSize is the max number of writes to send in a request
+	MaxBatchSize = 20
+	// RetryMaxBatchSize is the max number of writes to send in a retry request
+	RetryMaxBatchSize = 10
+	// MaxRetryAttempts is the max number of times to retry a write
+	MaxRetryAttempts = 10
+	// DefaultStartingMaximumOpsPerSecond is the starting max number of requests to the service per second
+	DefaultStartingMaximumOpsPerSecond = 500
+	// RateLimiterMultiplier is the amount to increase maximum ops (qps) every 5 minutes
+	RateLimiterMultiplier       = 1.5
+	rateLimiterMultiplierMillis = 5 * 60 * 1000 // 5 minutes in milliseconds
+	coolingPeriodMillis         = 2.0           // starting time to wait in between requests
+)
+
+type bulkWriterJob struct {
+	err      chan error           // send error responses to this channel
+	result   chan *pb.WriteResult // send the write results to this channel
+	write    *pb.Write            // the writes to apply to the database
+	attempts int                  // number of times this write has been attempted
+}
+
+type bulkWriterBatch struct {
+	bwr *pb.BatchWriteRequest // request to send to the service
+	bwj []bulkWriterJob       // corresponding jobs to return response
+}
+
+type bulkWriterRequestBatch struct {
+	isFlushing bool              // flag that indicates that BulkWriter needs to schedule a flush of the queue
+	bwb        []bulkWriterBatch // all the bulkWriterBatch objects to schedule
+}
+
+type bulkWriterStats int
+
+const (
+	writesProvided bulkWriterStats = iota
+	writesSent
+	writesReceived
+)
+
+// The BulkWriterStatus provides details about the BulkWriter, including the
+// number of writes processed, the number of requests sent to the service, and
+// the number of writes in the queue.
+type BulkWriterStatus struct {
```

Comment:
I did not look, but is this lib already instrumented with metrics? This seems like these should just be meterics on first glance.

Reply:
Interesting! Will take a look. 

This `Status()` is to allow the caller to wait until certain conditions are met--e.g. number of writes the caller sent is equal to the number provided, as an example. Rather than trying to interpret how the caller might want to manage their `BulkWriter`, I figured that I would give them the data to make decisions themselves.

As always, I'm open to alternative options!